### PR TITLE
[Spec] Add v1::AvgPool and v1::MaxPool

### PR DESCRIPTION
**Pull request by tsocha**
Rename and reorder attributes in `AvgPool` and `MaxPool` operators.
```
Shape m_window_shape                      -> m_kernel
Strides m_window_movement_strides         -> m_strides
Shape m_padding_below                     -> m_pads_begin
Shape m_padding_above                     -> m_pads_end
bool m_include_padding_in_avg_computation -> m_exclude_pad (logic is reversed Avg only)
PadType m_pad_type                        -> m_auto_pad 
bool m_ceil_mode                          -> RoudingType m_rounding_type
```

### CODE REVIEW

File: src/ngraph/op/avg_pool.cpp [hunk 1]
```diff
@@ -369,12 +371,328 @@ void op::AvgPool::generate_adjoints(autodiff::Adjoints& adjoints, const NodeVect
...
+
+void op::v1::AvgPoolBackprop::set_exclude_pad(bool exclude_pad)
+{
+    m_exclude_pad = exclude_pad;
+}
+
+shared_ptr<Node> op::v1::AvgPoolBackprop::copy_with_new_args(const NodeVector& new_args) const
+{
+    check_new_args_count(this, new_args);
+    AvgPoolBackprop* avpn = new AvgPoolBackprop(m_forward_arg_shape,
```
Comment (tomdol) [inline src/ngraph/op/avg_pool.cpp]: Is there any reason of not using `make_shared` here?

Comment (aprocter) [reply src/ngraph/op/avg_pool.cpp]: I think this is carried forward from the `v0` code. No reason AFAIK; I've noticed it a few times as I've been browsing the code but we've never gotten around to fixing it.

File: src/ngraph/op/avg_pool.cpp [hunk 2]
```diff
@@ -21,16 +21,17 @@
...
-op::AvgPool::AvgPool(const Output<Node>& arg,
-                     const Shape& window_shape,
-                     const Strides& window_movement_strides,
-                     const Shape& padding_below,
-                     const Shape& padding_above,
-                     bool include_padding_in_avg_computation,
-                     const PadType& pad_type,
-                     bool ceil_mode)
+// *** AvgPool OP SET 0 ***
+const string op::v0::AvgPool::type_name{"AvgPool"};
```
Comment (diyessi) [inline src/ngraph/op/avg_pool.cpp]: Check `abs.hpp` and `abs.cpp` and `changes.md` for the new way to do this.

File: src/ngraph/op/avg_pool.cpp [hunk 3]
```diff
@@ -369,12 +371,328 @@ void op::AvgPool::generate_adjoints(autodiff::Adjoints& adjoints, const NodeVect
...
+                                                         m_window_movement_strides,
+                                                         m_padding_below,
+                                                         m_padding_above,
+                                                         m_include_padding_in_avg_computation);
     adjoints.add_delta(operand, backprop);
 }
+
+// *** AvgPool OP SET 1 ***
+const string op::v1::AvgPool::type_name{"AvgPool"};
+
```
Comment (diyessi) [inline src/ngraph/op/avg_pool.cpp]: This one also needs to be upgraded.

File: src/ngraph/op/avg_pool.cpp (hunk 1 above)
Comment (diyessi) [reply src/ngraph/op/avg_pool.cpp]: Please fix it both places.

File: src/ngraph/serializer.cpp [hunk 4]
```diff
@@ -1418,66 +1460,101 @@ shared_ptr<Node> JSONDeserializer::deserialize_node(json node_js)
...
-                node = make_shared<op::MaxPool>(args[0],
-                                                window_shape,
-                                                window_movement_strides,
-                                                padding_below,
-                                                padding_above,
-                                                pad_type);
+                auto window_shape = node_js.at("window_shape").get<vector<size_t>>();
+                auto window_movement_strides =
+                    node_js.at("window_movement_strides").get<vector<size_t>>();
+                // For backwards compatibility, both (but not just one) of the padding_ fields may
```
Comment (aprocter) [inline src/ngraph/serializer.cpp]: Nice to be able to get rid of this nonsense in v1 :)

File: src/ngraph/op/avg_pool.hpp [hunk 5]
```diff
@@ -23,191 +23,353 @@ namespace ngraph
...
+                /// `[n]`
+                /// \param window_movement_strides The window movement strides.<br>
+                /// `[n]`
+                /// \param padding_below The below-padding shape.<br>
+                /// `[n]`
+                /// \param padding_above The above-padding shape.<br>
+                /// `[n]`
+                /// \param include_padding_in_avg_computation If true then averages include padding
+                ///  elements, each treated as the number zero.  If false, padding elements are
+                ///  entirely
```
[the remark this answers is not in the document]
Comment (tsocha) [reply src/ngraph/op/avg_pool.hpp]: Done.

**Comment by tsocha** [comment]
> @postrational 
> Looks good. Please add some unit tests.
Done.

**Pull request merged**